client/cli/src/commands/vanity.rs

```diff
@@ -16,77 +16,92 @@
+use structopt::StructOpt;
+use rand::{rngs::OsRng, RngCore};
+use sp_runtime::traits::IdentifyAccount;
+
+/// The `vanity` command
+#[derive(Debug, StructOpt)]
+#[structopt(
+	name = "vanity",
+	about = "Generate a seed that provides a vanity address"
+)]
+pub struct VanityCmd {
+	/// Number of keys to generate
+	#[structopt(long, short)]
+	number: String,
+
+	/// Desired pattern
+	#[structopt(long)]
+	pattern: Option<String>,
+
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub shared_params: SharedParams,
+
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub network_scheme: NetworkSchemeFlag,
+
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub output_scheme: OutputTypeFlag,
+
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub crypto_scheme: CryptoSchemeFlag,
 }
 
-fn next_seed(seed: &mut [u8]) {
-	for i in 0..seed.len() {
-		match seed[i] {
-			255 => {
-				seed[i] = 0;
-			}
-			_ => {
-				seed[i] += 1;
-				break;
-			}
-		}
+impl VanityCmd {
+	/// Run the command
+	pub fn run(&self) -> error::Result<()> {
+		let desired: &str = self.pattern.as_ref()
+			.map(String::as_str)
+			.unwrap_or("");
+		let formated_seed = with_crypto_scheme!(self.crypto_scheme.scheme, generate_key(desired))?;
+		use utils::print_from_uri;
+		with_crypto_scheme!(
+			self.crypto_scheme.scheme,
+			print_from_uri(
+				&formated_seed,
+				None,
+				self.network_scheme.network.clone(),
+				self.output_scheme.output_type.clone()
+			)
+		);
+		Ok(())
 	}
 }
 
-/// A structure used to carry both Pair and seed.
-/// This should usually NOT been used. If unsure, use Pair.
-pub(super) struct KeyPair<C: Crypto> {
-	pub pair: C::Pair,
-	pub seed: <C::Pair as Pair>::Seed,
-	pub score: usize,
-}
-
-/// Calculate the score of a key based on the desired
-/// input.
-fn calculate_score(_desired: &str, key: &str) -> usize {
-	for truncate in 0.._desired.len() {
-		let snip_size = _desired.len() - truncate;
-		let truncated = &_desired[0..snip_size];
-		if let Some(pos) = key.find(truncated) {
-			return (47 - pos) + (snip_size * 48);
-		}
+impl CliConfiguration for VanityCmd {
+	fn shared_params(&self) -> &SharedParams {
+		&self.shared_params
 	}
-	0
-}
-
-/// Validate whether the char is allowed to be used in base58.
-/// num 0, lower l, upper I and O are not allowed.
-fn validate_base58(c :char) -> bool {
-	c.is_alphanumeric() && !"0lIO".contains(c)
 }
 
-pub(super) fn generate_key<C: Crypto>(desired: &str) -> Result<KeyPair<C>, &'static str> where
-		PublicOf<C>: PublicT,
+/// genertae a key based on given pattern
+fn generate_key<Pair>(desired: &str) -> Result<String, &'static str>
+	where
+		Pair: sp_core::Pair,
+		Pair::Public: IdentifyAccount,
+		<Pair::Public as IdentifyAccount>::AccountId: Ss58Codec,
 {
 	if desired.is_empty() {
 		return Err("Pattern must not be empty");
 	}
 
-	if !desired.chars().all(validate_base58) {
-		return Err("Pattern can only contains valid characters in base58 \
-			(all alphanumeric except for 0, l, I and O)");
-	}
```

Comment:
In fact, what would be better than using a `String` for this would be to use a wrapper type that performs these checks in `FromStr`, so that it fails in the CLI interface and not here.